src/AI/ModularAI.java

```diff
@@ -0,0 +1,181 @@
+    {
+      for( int y = 0; y < gameMap.mapHeight; ++y )
+      {
+        XYCoord loc = new XYCoord(x, y);
+        if( gameMap.getLocation(loc).isCaptureable() && myCo.isEnemy(gameMap.getLocation(loc).getOwner()) )
+        {
+          unownedProperties.add(loc);
+        }
+      }
+    }
+
+    for( AIModule phase : aiPhases )
+    {
+      phase.initTurn(gameMap);
+    }
+  }
+
+  protected void log(String message)
+  {
+    System.out.println(message);
+    logger.append(message).append('\n');
+  }
+
+  @Override
+  public GameAction getNextAction(GameMap gameMap)
+  {
+    ArrayList<Unit> eligibleUnits = new ArrayList<Unit>();
+    for( Unit unit : myCo.units )
+    {
+      if( unit.isTurnOver || !gameMap.isLocationValid(unit.x, unit.y) )
+        continue; // No actions for units that are stale or out of bounds.
+      eligibleUnits.add(unit);
+    }
+
+    GameAction nextAction = null;
+    PriorityQueue<Unit> unitQueue = new PriorityQueue<Unit>(13, unitOrderSetter);
+
+    // Each module gets a shot at the whole unit queue, in order.
+    // If no action's found, we're done
+    for( AIModule phase : aiPhases )
+    {
+      unitQueue.clear();
+      unitQueue.addAll(eligibleUnits);
+
+      nextAction = phase.getNextAction(unitQueue, gameMap);
+      if( null != nextAction )
+        break;
+    }
+    log(String.format("Action: %s", nextAction));
+    return nextAction;
+  }
+
+  public static interface AIModule extends Serializable
+  {
+    public GameAction getNextAction(PriorityQueue<Unit> unitQueue, GameMap map);
+    public default void initTurn(GameMap gameMap) {}
+  }
+
+  public static class PowerActivator implements AIModule
+  {
+    private static final long serialVersionUID = 1L;
+    public Commander myCo;
+    public final int aiPhase;
+
+    public PowerActivator(Commander co, int phase)
+    {
+      myCo = co;
+      aiPhase = phase;
+    }
+
+    public boolean checked = false;
+    @Override
+    public void initTurn(GameMap gameMap) { checked = false; }
+    @Override
+    public GameAction getNextAction(PriorityQueue<Unit> unitQueue, GameMap map)
+    {
+      if( checked )
+        return null;
+      checked = true;
+
+      CommanderAbility retVal = AIUtils.queueCromulentAbility(null, myCo, aiPhase);
+      if( null != retVal )
+        return new GameAction.AbilityAction(retVal);
+      return null;
+    }
+  }
+
+  public static class CaptureFinisher extends UnitActionFinder
+  {
+    private static final long serialVersionUID = 1L;
+    public CaptureFinisher(Commander co, ModularAI ai)
+    {
+      super(co, ai);
+    }
+
+    @Override
+    public GameAction getUnitAction(Unit unit, GameMap map)
+    {
+      if( unit.getCaptureProgress() > 0 )
+      {
+        XYCoord position = new XYCoord(unit.x, unit.y);
+        ai.unownedProperties.remove(position);
+        return new CaptureLifecycle.CaptureAction(map, unit, Utils.findShortestPath(unit, position, map));
```

Comment:
This is a pretty decent "default" capture-continuation behavior, but it doesn't really accommodate situations where you don't want to; e.g. a high-value target is adjacent and only reachable from the spot being captured, or the capturing unit is about to be stomped by nearby enemies and won't be able to finish anyway, or the unit has already been damaged and a healthy friend is nearby who can complete the capture faster.

One of the main concerns I have with this whole approach is that although it makes it easier to get a new AI running with basic functionality, it hides a lot of the complexity and reduces the problem to trying to find the right building blocks, when a successful strategy will necessarily use a more (oh boy I get to use this in a sentence) holistic approach.

Reply:
> This is a pretty decent "default" capture-continuation behavior, but it doesn't really accommodate situations where you don't want to; e.g. a high-value target is adjacent and only reachable from the spot being captured, or the capturing unit is about to be stomped by nearby enemies and won't be able to finish anyway, or the unit has already been damaged and a healthy friend is nearby who can complete the capture faster.

All of this is true, and a great argument for... putting all that cool logic into higher-priority modules.
Or you can roll your own.
It'd be just as valid to put this at the end of your module list, so long as you account for the fact that not all your units will have finished any applicable captures.

> One of the main concerns I have with this whole approach is that although it makes it easier to get a new AI running with basic functionality, it hides a lot of the complexity

Isn't that... the *whole point* of an abstraction?
This AI base gives you:
a) a very concise and idiomatic way of specifying what sorts of actions your AI will take, and in what order they will be considered/executed.
b) free management of a few key things like finding the "highest priority" unit to act.

You don't really lose any control.

> ...and reduces the problem to trying to find the right building blocks

Not really? It just makes it easier to put them together.

> ...when a successful strategy will necessarily use a more (oh boy I get to use this in a sentence) holistic approach.

How is "defining what I'm gonna do and what my priorities are" not a holistic approach?

Reply:
> putting all that cool logic into higher-priority modules

Yeah, the priority element is pretty interesting, especially if it allows module A to pass, let module B to do something, and then pick an action the next time around.

> Isn't that... the whole point of an abstraction?

Well, yes. And no. But mostly yes. It's good when you don't need that complexity to solve the problem effectively. It's bad when it hides or obscures features that would be useful.
As you say though, an AI builder can simply swap out modules or provide higher-priority ones to handle stuff they care about.

>  How is "defining what I'm gonna do and what my priorities are" not a holistic approach?

Depends on how you decide what you are going to do. Holistic just means to consider the whole problem space when choosing an action. `CaptureFinisher`, for example, currently doesn't look out for threats to itself (though it could if designed that way). It's not very holistic, (and maybe it doesn't need to be, to be effective).

My comment just meant that this modular strategy encourages throwing together a bunch of myopic building blocks that don't know anything about what the others are doing, and might well end up working at cross-purposes if someone isn't careful. That doesn't mean it's a bad approach, it just means care should be taken, which is just... true for any attempt to create a non-trivial AI.

Reply:
> Yeah, the priority element is pretty interesting, especially if it allows module A to pass, let module B to do something, and then pick an action the next time around.

It totally does. And it makes your code *prettier* (IMO) when you use it, instead of uglier, now.

> As you say though, an AI builder can simply swap out modules or provide higher-priority ones to handle stuff they care about.

...and this scheme provides a concise way to define your order of operations/priority, so you can move stuff around without huge deltas.
You could do that anyway by factoring everything into methods, but this is easier to look at than that would be IMO.

> Depends on how you decide what you are going to do. Holistic just means to consider the whole problem space when choosing an action.

That's fair; this design is relatively ill-suited for an AI that considers/scores all potential moves before choosing one.
It would still gain a *little* from the inversion-of-control bit, but yeah.
A reasonable extension might be to have a variant that polls an action (and a score for that action) from all modules and picks the highest... but defining the scoring metric sounds pretty hard lol

Such an AI would probably also lag quite a bit more than Wally does, and that's saying something lol

> `CaptureFinisher`, for example, currently doesn't look out for threats to itself (though it could if designed that way). It's not very holistic, (and maybe it doesn't need to be, to be effective).

Yeah. That's fair.
You can get pretty holistic through the combination of fairly myopic parts, though, so long as they're properly designed.

> My comment just meant that this modular strategy encourages throwing together a bunch of myopic building blocks that don't know anything about what the others are doing, and might well end up working at cross-purposes if someone isn't careful. That doesn't mean it's a bad approach, it just means care should be taken, which is just... true for any attempt to create a non-trivial AI.

The way I look at it is it just makes it *easier* to slap your building blocks together, rather than encouraging you to do so haphazardly.
You'll note I also put in a fair amount of pre/post-conditions in comments along the line of "I expect to be called after X has done stuff this round"

This also nudges the door open for possibly letting COs define AI routines for activating/maximizing benefit from their powers, which I'm kind of a fan of - this would probably require more wiring to let the AI tell its modules the sorts of things it values, so e.g. Meridian's module can transform Tanks to Arty in places the AI prefers before COP-ing.

Reply:
> You'll note I also put in a fair amount of pre/post-conditions in comments along the line of "I expect to be called after X has done stuff this round"

I did not see that. I don't see many comments on the AI modules at all. I'm also not sure about having implicit dependencies between modules anyway, but eh, I guess that's up to the AI designer.

> letting COs define AI routines

That sounds like a fun/scary idea. Meridian will finally be relevant!

Reply:
> I did not see that. I don't see many comments on the AI modules at all. I'm also not sure about having implicit dependencies between modules anyway, but eh, I guess that's up to the AI designer.

Hmm. I've definitely had a bit of a shift in philosophy towards comments.
If I can't say something that meaningfully summarizes a fair amount of logic nor something that expresses design intent significantly more clearly than the code... adding another comment to maintain seems like make-work.
Example of this:
```java
  /** Functions as working memory to prevent eviction cycles */
  private transient Set<Unit> evictionStack;
```
I don't repeat the exact "what" of it, since that'd be kind of redundant. Instead, I say why I'm tracking it at all.
I guess I could go into more detail, but I'm not sure that I'd be clearer than just reading the code, and keeping two copies of the algo (one which the compiler doesn't care about) seems silly.

I did at least note `For use after unit building is complete` a few places. I coulda sworn I added in a thing about the threat map; I should do that.

I feel like you're not gonna get away with having no dependency between modules at all, unless you really like duplicating calculations or something.

> > letting COs define AI routines
> 
> That sounds like a fun/scary idea. Meridian will finally be relevant!

Hah. Indeed.
Pretty sure I've even brought it up before a few times, lol

Reply:
Yeah, using comments for the "why"--the intent--is probably the way to go. Which your eviction example follows, but not your "use after unit building" example. Why should the module be used only after producing units?

> I feel like you're not gonna get away with having no dependency between modules at all

I mean, you could get clever and accept that data in the interface, and only recalculate it if it isn't given so the module can be generically used whether or not the AI optimizes it properly. But that would be potentially complex to implement, and may or may not be worth it.

Reply:
> Why should the module be used only after producing units?

Because it doesn't consider whether the unit's moving onto a production building :P 

> I mean, you could get clever and accept that data in the interface, and only recalculate it if it isn't given so the module can be generically used whether or not the AI optimizes it properly. But that would be potentially complex to implement, and may or may not be worth it.

But now the interface is no longer generic - it has to be specialized to the module in question.
...and if it's unspecialized data, you'd have to just assume what it is and cast it to whatever you think your data is and hope it works.